torchgeo/trainers/simclr.py

```diff
@@ -0,0 +1,242 @@
+        layers: int = 3,
+        hidden_dim: Optional[int] = None,
+        output_dim: Optional[int] = None,
+        lr: float = 4.8,
+        weight_decay: float = 1e-4,
+        max_epochs: int = 100,
+        temperature: float = 0.07,
+        memory_bank_size: int = 64000,
+        gather_distributed: bool = False,
+        augmentations: nn.Module = AUG,
+    ) -> None:
+        """Initialize a new SimCLRTask instance.
+
+        Args:
+            model: Name of the timm model to use.
+            weights: Either a weight enum, the string representation of a weight enum,
+                True for ImageNet weights, False or None for random weights,
+                or the path to a saved model state dict.
+            in_channels: Number of input channels to model.
+            version: Version of SimCLR, 1--2.
+            layers: Number of layers in projection head (2 for v1 or 3+ for v2).
+            hidden_dim: Number of hidden dimensions in projection head
+                (defaults to output dimension of model).
+            output_dim: Number of output dimensions in projection head
+                (defaults to output dimension of model).
+            lr: Learning rate (0.3 x batch_size / 256 is recommended).
+            weight_decay: Weight decay coefficient (1e-6 for v1 or 1e-4 for v2).
+            max_epochs: Maximum number of epochs to train for.
+            temperature: Temperature used in NT-Xent loss.
+            memory_bank_size: Size of memory bank (0 for v1 or 64K for v2).
+            gather_distributed: Gather negatives from all GPUs during distributed
+                training (ignored if memory_bank_size > 0).
+            augmentations: Data augmentation.
+        """
+        super().__init__()
+
+        self.save_hyperparameters(ignore=["augmentations"])
+        self.augmentations = augmentations
+
+        # Create backbone
+        self.backbone = timm.create_model(
+            model, in_chans=in_channels, pretrained=weights is True
+        )
+
+        # Load weights
+        if weights and weights is not True:
+            if isinstance(weights, WeightsEnum):
+                state_dict = weights.get_state_dict(progress=True)
+            elif os.path.exists(weights):
+                _, state_dict = utils.extract_backbone(weights)
+            else:
+                state_dict = get_weight(weights).get_state_dict(progress=True)
+            self.backbone = utils.load_state_dict(self.backbone, state_dict)
+
+        # Create projection head
+        input_dim = list(self.backbone.named_children())[-1][1].out_features
+        if hidden_dim is None:
+            hidden_dim = input_dim
+        if output_dim is None:
+            output_dim = input_dim
+
+        self.projection_head = SimCLRProjectionHead(
+            input_dim, hidden_dim, output_dim, layers
+        )
+
+        # Define loss function
+        self.criterion = NTXentLoss(temperature, memory_bank_size, gather_distributed)
```

Comment:
v1 doesn't use a memory bank.  We should check if a user set `version=1` and set `memory_bank_size` to None

Reply:
I guess it depends on how flexible we want to be. We could force layers, weight_decay, and memory_bank_size to match the version, although this limits flexibility if you want to experiment with one feature but not another. At the moment, we aren't actually using version for anything, but we could.

Reply:
We should at least put a note in the docstring. My concern is based on user blindly assume changing `version=1` would literally mean that they are using SimCLR v1 when in fact they still need to change the `weight decay` and `memory_bank` args which seems to defeat the purpose of even having a `version` arg.

Reply:
Should we compromise on a warning message?

Reply:
I'm good with that!